django/db/models/query.py

```diff
@@ -470,6 +470,41 @@ def bulk_create(self, objs, batch_size=None):
 
         return objs
 
+    def bulk_save(self, objs, fields, batch_size=None):
+        """
+        Update the given fields in each of the given objects in the database.
+        Do *not* call save() on each of the instances.
+        """
+        assert batch_size is None or batch_size > 0
+        if not fields:
+            raise ValueError('Specific field names must be given to bulk_save')
+        objs = list(objs)
+        if not all(obj.pk for obj in objs):
+            raise ValueError('All objects must have a primary key set')
+        if not objs:
+            return
+        fields = list(fields)
+        ops = connections[self.db].ops
+        # We use PK twice in the resulting update query, once in the filter
+        # and once in the WHEN.
+        batch_size = (batch_size or max(ops.bulk_batch_size(['pk', 'pk'] + fields, objs), 1))
+        fields = {
+            field: self.model._meta.get_field(field)
+            for field in fields
+        }
+        with transaction.atomic(using=self.db, savepoint=False):
+            for batch_objs in [objs[i:i + batch_size] for i in range(0, len(objs), batch_size)]:
+                pks = [obj.pk for obj in batch_objs]
+                update_kwargs = {
```

Comment:
My approach to solving this was slightly different: http://schinckel.net/2017/06/30/django-bulk_update-without-upsert/

Basically, instead of a CASE WHEN for each field/row, it puts the data into a `VALUES` list, and joins that in.

     UPDATE my_table
        SET foo = upsert_source.column2,
            column2 = upsert_source.column3
       FROM (
         VALUES (...), (...)
       ) AS upsert_source
      WHERE upsert_source.column1 = my_table.id;

I'd be interested to see how these two different approaches compare in terms of performance.

The big advantage of your approach is that it uses the ORM "more".

Reply:
How widely supported is that method? I know Postgres supports it and I was intending to work on adding database specific methods at a later point (mainly for performance reasons more than anything), but as far as I can tell the currently implemented method will run on any database.

(Github added this as a separate comment rather than replying to this thread, so I'm copy+pasting it here)